File: python/0098验证二叉搜索树.py
from binarytree import TreeNode

# BST的中序遍历节点数值是从小到大. 
def isValidBST(self, root: TreeNode) -> bool:
    cur_max = -float("INF")
    def _isValidBST(root: TreeNode) -> bool: 
        nonlocal cur_max
        
        if not root: 
            return True
        
        is_left_valid = _isValidBST(root.left)

        if cur_max >= root.val: 
            return False
        cur_max = root.val
        
        is_right_valid = _isValidBST(root.right)
        
        return is_left_valid and is_right_valid
    return _isValidBST(root)
        

File: cpp/0098验证二叉搜索树.cpp
#include"mybitree.h"
#include<stack>
#include<vector>
#include<algorithm>

bool isValidBST(TreeNode *root)
{
    stack<TreeNode *> nodes;
    nodes.push(root);
    vector<TreeNode *> memo;
    TreeNode *pre = nullptr;

    while (!nodes.empty())
    {
        TreeNode *crt = nodes.top();

        if(crt->left && (find(memo.begin(),memo.end(),crt->left)==memo.end()))
            nodes.push(crt->left);
        
        else
        {
            nodes.pop();
            memo.push_back(crt);
            if (pre == nullptr || pre->val < crt->val)
                pre = crt;
            else
                return false;
            if(crt->right)
                nodes.push(crt->right);
        }

    }

    return true;
}